auth/api/iam/error.go

```diff
@@ -0,0 +1,125 @@
+	ServerError ErrorCode = "server_error"
+)
+
+// Make sure the error implements core.HTTPStatusCodeError, so the HTTP request logger can log the correct status code.
+var _ core.HTTPStatusCodeError = OAuth2Error{}
+
+// OAuth2Error is an OAuth2 error that signals the error was (probably) caused by the client (e.g. bad request),
+// or that the client can recover from the error (e.g. retry).
+type OAuth2Error struct {
+	// Code is the error code as defined by the OAuth2 specification.
+	Code ErrorCode `json:"error"`
+	// Description is a human-readable ASCII [USASCII] text providing additional information, used to assist the client developer in understanding the error that occurred.
+	Description string `json:"error_description,omitempty"`
+	// InternalError is the underlying error, may be omitted. It is not intended to be returned to the client, only to be logged.
+	InternalError error `json:"-"`
+	// RedirectURI is the redirect URI that should be used to redirect the client to, in case the user-agent is a browser.
+	// It should not be set if the user-agent is not a browser, or there is no redirect_uri (because the request was malformed), this field is empty.
+	// When the field is set, the user-agent is redirected to the specified URI with the error code and description as query parameters.
+	// If it's not set, the error code and description are returned in the response body (plain text or JSON).
+	RedirectURI string `json:"-"`
+}
+
+// StatusCode returns the HTTP status code to be returned to the client, in case the user-agent can't be redirected with HTTP 302 - Found.
+func (e OAuth2Error) StatusCode() int {
+	switch e.Code {
+	case ServerError:
+		return http.StatusInternalServerError
+	default:
+		return http.StatusBadRequest
+	}
+}
+
+// OAuth2Error returns the error message, which is either the underlying error or the code if there is no underlying error
+func (e OAuth2Error) Error() string {
+	var parts []string
+	parts = append(parts, string(e.Code))
+	if e.InternalError != nil {
+		parts = append(parts, e.InternalError.Error())
+	}
+	if e.Description != "" {
+		parts = append(parts, e.Description)
+	}
+	return strings.Join(parts, " - ")
+}
+
+type oauth2ErrorWriter struct{}
+
+func (p oauth2ErrorWriter) Write(echoContext echo.Context, _ int, _ string, err error) error {
+	var oauthErr OAuth2Error
+	if !errors.As(err, &oauthErr) {
+		// Internal error, wrap it in an OAuth2 error
+		oauthErr = OAuth2Error{
+			Code:          ServerError,
+			InternalError: err,
+		}
+	}
+	if oauthErr.Code == "" {
+		// Somebody forgot to set a code
+		oauthErr.Code = ServerError
+	}
+	redirectURI, _ := url.Parse(oauthErr.RedirectURI)
+	if oauthErr.RedirectURI == "" || redirectURI == nil {
+		// Can't redirect the user-agent back, render error as JSON or plain text (depending on content-type)
+		contentType := echoContext.Request().Header.Get("Content-Type")
```

Comment:
`Content-Type` or `Accept` header?

Reply:
At first, I had both. But that introduced unused code paths (`Accept` header being `application/json`), which adds complexity and no value. So given the assumption that a browser will never send `application/json` as content-type (maybe with a form post with enctype set to that value, but that would be very, very weird?), and API clients will use JSON, I just settles on `Content-Type = application/json` for now.

When proven otherwise in future, we can easily adapt this.